jenkins/agent-ctl.sh

```diff
@@ -0,0 +1,315 @@
+fi
+
+while true; do
+  case "${1:-}" in
+    --real)
+      REAL=yes
+      shift
+      ;;
+    --pr)
+      PR=yes
+      shift
+      ;;
+    *)
+      break
+      ;;
+  esac
+done
+
+INSTANCE="${1}"
+shift
+if [[ -n "${PR}" ]]; then
+  echo 'Talking to PR jenkins'
+  MASTER='pull-jenkins-master'
+else
+  MASTER='jenkins-master'
+fi
+
+if [[ "${INSTANCE}" =~ light- ]]; then
+  KIND='light'
+elif [[ "${INSTANCE}" =~ heavy- ]]; then
+  KIND='heavy'
+elif [[ "${INSTANCE}" =~ pr- ]]; then
+  KIND='pr'
+else
+  KIND=
+fi
+
+case "${KIND}" in
+  light)
+    # Current experiment:
+    # 20 executors, n1-highmem-16, 500G pd-standard
+    # Results:
+    # 5.31 cores, 70G ram, <500 write IOPs, <50MB/s write
+    # Alernate experiment:
+    # 2 executors, n1-highmem-2, 100G pd-standard
+    # Results:
+    # 0.2 cores, 1.3G ram, low IOPs (200 IOP spikes), low write (30MB/s spikes)
+    DISK_SIZE='500GB'
+    DISK_TYPE='pd-standard'
+    MACHINE_TYPE='n1-highmem-16'
+    ;;
+  heavy)
+    # TODO(fejta): optimize these values
+    DISK_SIZE='200GB'
+    DISK_TYPE='pd-standard'
+    MACHINE_TYPE='n1-standard-16'
+    ;;
+  pr)
+    # TODO(fejta): optimize these values
+    DISK_SIZE='500GB'
+    DISK_TYPE='pd-ssd'
+    MACHINE_TYPE='n1-highmem-32'
+    ;;
+  *)
+    ;;
+esac
+
+if [[ -z "${REAL}" ]]; then
+  DISK_SIZE='200GB'
+  DISK_TYPE='pd-standard'
+  MACHINE_TYPE='n1-standard-1'
+  read -p "Using ${MACHINE_TYPE} for testing. Continue [Y/n]: " ans
+  if [[ ! "${ans}" =~ '^[yY]' ]]; then
+    echo 'Add --real'
+    exit 1
+  fi
+fi
+
+check-kind() {
+  if [[ -z "${KIND}" ]]; then
+    echo "${INSTANCE} does not contain light-|heavy-|pr-"
+    exit 1
+  fi
+}
+
+auto-agent() {
+  echo "Automatically creating ${INSTANCE}..."
+  check-kind
+  detatch-agent
+  delete-agent
+  create-agent
+  update-agent
+  reboot-agent
+  attach-agent
+}
+
+tunnel-to-master() {
+  for i in {1..10}; do
+    if sudo netstat -anp | grep 8080 > /dev/null 2>&1 ; then
+      sleep 1
+    else
+      break
+    fi
+  done
+  sudo netstat -anp | grep 8080 && echo "8080 already used" && exit 1 \
+    || echo "Tunneling to ${MASTER}..."
+  gcloud compute ssh "${MASTER}" --ssh-flag='-L8080:localhost:8080' sleep 5 &
+  for i in {1..10}; do
+    if sudo netstat -anp | grep 8080 > /dev/null 2>&1 ; then
+      break
+    fi
+    sleep 1
+  done
+  sudo netstat -anp | grep 8080 > /dev/null 2>&1 || exit 1
+}
+
+
+master-change() {
+  tunnel-to-master
+  cmd="${1}"
+  ini="${HOME}/jenkins-master-creds.ini"  # /user/<user>/configure && show api token
+  if [[ ! -f "${ini}" ]]; then
+    echo "Missing config: ${ini}"
+    exit 1
+  fi
+  python "$(dirname "${0}")/attach_agent.py" "${cmd}" "${INSTANCE}" "${KIND}" "${ini}" "${MASTER}"
+  echo 'Waiting for tunnel to close...'
+  wait
+}
+
+
+detatch-agent() {
+  echo "Detatching ${INSTANCE}..."
+  master-change delete
+}
+
+
+attach-agent() {
+  gcloud compute ssh "${INSTANCE}" -- gcloud compute instances list "--filter=name=${INSTANCE}"
```

Comment:
ah. can you add a comment indicating that's what it's doing, then? or maybe print that if the check fails?